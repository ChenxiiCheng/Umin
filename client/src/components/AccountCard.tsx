import React, { useState, useEffect } from 'react'
import { Link, useNavigate, RouteComponentProps, WindowLocation } from '@reach/router';
import { useMutation } from '@apollo/client';
import { Box, Text, Heading, InputGroup, InputLeftAddon, Input, Button, FormControl, useColorModeValue, InputRightElement, useToast } from '@chakra-ui/core';
import {FaRegEnvelope, FaRegEye, FaRegEyeSlash, FaRegStar} from 'react-icons/fa'

import { LOGIN, REGISTER } from '../graphql/user';

interface IProps extends RouteComponentProps {
  type: 'register' | 'login',
}

interface LoginPayload {
  email: string
  nickname?: string
  password: string
}

export const AccountCard: React.FC<IProps> = ({ type, location }) => {
  const navigate = useNavigate()
  const toast = useToast()
  const [login, { error: loginMutaionError, loading: loginLoading }] = useMutation(LOGIN)
  const [register, {error: registerMutaionError, loading: registerLoading}] = useMutation(REGISTER)
  const cardBgColor = useColorModeValue('white', 'rgba(26, 33, 46, 0.8)')
  const textColor = useColorModeValue('#1A202C', 'white')
  const boxShadowMode = useColorModeValue('md', '2px 4px 6px -1px rgba(250,255,255,0.1), 0px 2px 4px -1px rgba(255,255,255,0.1)')
  const [email, setEmail] = useState('');
  const [nickname, setNickname] = useState('');
  const [password, setPassword] = useState('');
  const [emailError, setEmailError] = useState(false)
  const [passwordError, setPasswordError] = useState(false)
  const [nicknameError, setNicknameError] = useState(false)
  const [showPassword, setShowPassword] = useState(false)

  // * listen graphql response error
  useEffect(() => {
    if (typeof loginMutaionError !== "undefined" || typeof registerMutaionError !== "undefined") {
      const errorMsg = (loginMutaionError ? loginMutaionError?.graphQLErrors[0].message : "") || (registerMutaionError ? registerMutaionError?.graphQLErrors[0].message : "")
      toast({
        title: "Failed",
        status: "error",
        description: errorMsg,
        duration: 2500,
        position: "top",
        isClosable: true
      })
    }
  }, [loginMutaionError, registerMutaionError])

  const handleLoginSubmit = (e: React.FormEvent) => {
    e.preventDefault();

    let emailError = false, passwordError = false, nicknameError = false
    if (!email || !email.includes('@')) {
      emailError = true
      setEmailError(true)
      toast({
        title: "Error",
        description: "Please enter valid email address",
        status: "error",
        duration: 3000,
        position: "top",
        isClosable: true
      })
    }
    if (!password) {
      passwordError = true
      setPasswordError(true)
      toast({
        title: "Error",
        description: "Please enter correct password",
        status: "error",
        duration: 3000,
        position: "top",
        isClosable: true
      })
    }
    if (type === "register" && (!nickname || nickname.length < 4)) {
      nicknameError = true
      setNicknameError(true)
      toast({
        title: "Warning",
        description: "Minimum length of nickname is 4",
        status: "warning",
        duration: 3000,
        position: "top",
        isClosable: true
      })
    }

    const payload: LoginPayload = {
      email: email,
      password: password
    }

    if (nickname) {
      payload.nickname = nickname
    }

    // * call login or register backend endpoint
    if (!emailError && !passwordError && !nicknameError) {
      if (type === "login") {
        login({ variables: { user: payload } }).then(() => {
          navigate("/")
          toast({
            title: "Login Success",
            status: "success",
            duration: 2500,
            position: "top",
            isClosable: true
          })
        }).catch(() => {
          // * 想获取后端具体返回的错误信息，use exposed error state
          navigate("/login")
        })
      } else {
        register({ variables: { user: payload } }).then((res) => {
          navigate("/login")
          toast({
            title: "Register Success",
            status: "success",
            description: "Registered success, now please login in.",
            duration: 2500,
            position: "top",
            isClosable: true
          })
        }).catch(() => {
          navigate("/register")
        })
      }
    }
  }

  return (
    <form onSubmit={handleLoginSubmit}>
      <Box
        display="flex"
        flexDir="column"
        justifyContent="space-between"
        alignItems="center"
        shadow={boxShadowMode}
        borderWidth="1px"
        borderRadius="md"
        p={10}
        pt={9}
        minW={530}
        minH={type === 'login' ? 400 : 450}
        maxW={600}
        maxH={600}
        color={textColor}
        background={cardBgColor}
        fontSize="lg"
        fontWeight="450"
      >
        <Heading as="h2" size="xl">
          {type === "login" ? "Sign In" : "Sign Up"}
        </Heading>
        <Text color={textColor} _hover={{color: "gray.500"}}>
          {type === "login" ? 
            <Link to="/register">😱 Do not have an account?</Link>
          : <Link to="/login">🔥 Already have an account?</Link>}
        </Text>
        <InputGroup mt={3}>
          <InputLeftAddon
            pointerEvents="none"
            children={<FaRegEnvelope color="gray.300" />}
          />
          <FormControl isRequired>
            <Input
              type="text"
              pl="20px"
              focusBorderColor="gray.300"
              isInvalid={emailError === true ? true: false}
              errorBorderColor={emailError === true ? "red.300" : ""}
              placeholder="Email"
              value={email}
              onChange={(event) => setEmail(event.currentTarget.value)}
            />
          </FormControl>
        </InputGroup>
        {/* COMMENT: regster form has one extra input - Nickname */}
        {type === 'register' ? (
          <InputGroup>
            <InputLeftAddon
              pointerEvents="none"
              children={<FaRegStar color="gray.300" />}
            />
            <FormControl isRequired>
              <Input
                type="text"
                pl="20px"
                focusBorderColor="gray.300"
                isInvalid={nicknameError === true ? true: false}
                errorBorderColor={nicknameError === true ? "red.300" : ""}
                placeholder="Nickname"
                value={nickname}
                onChange={(event) => setNickname(event.currentTarget.value)}
              />
            </FormControl>
          </InputGroup>
        ) : null}
        <InputGroup>
          <InputLeftAddon
            pointerEvents="none"
            children={showPassword ? <FaRegEye color="gray.300" /> : <FaRegEyeSlash color="gray.300" />}
          />
          <FormControl isRequired>
            <Input
              type={showPassword ? "text" : "password"}
              focusBorderColor="gray.300"
              pl="20px"
              isInvalid={passwordError === true ? true: false}
              errorBorderColor={passwordError === true ? "red.300" : ""}
              placeholder="Password"
              value={password}
              onChange={(event) => setPassword(event.currentTarget.value)}
            />
          </FormControl>
          <InputRightElement>
            <Button
              h="2rem"
              w="3rem"
              size="md"
              mr={2}
              onClick={() => setShowPassword(!showPassword)}
            >
              <Text color="gray.500">{showPassword ? "Hide" : "Show"}</Text>
            </Button>
          </InputRightElement>
        </InputGroup>
        <Button color={textColor} w="100%" type="submit" isLoading={loginLoading || registerLoading}>
          {type === "login" ? "Sign In" : "Sign Up"}
        </Button>
      </Box> 
    </form>
  )
}
