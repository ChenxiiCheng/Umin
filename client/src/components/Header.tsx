import React, {useState, useRef} from 'react'
import { Text, Link, Button, HStack, Box, Avatar, AvatarBadge, Image, useColorModeValue, Menu, MenuButton, MenuList, MenuItem, AlertDialog, AlertDialogOverlay, AlertDialogContent, AlertDialogHeader, AlertDialogBody, AlertDialogFooter } from '@chakra-ui/core'
import { MdBookmarkBorder, MdControlPoint } from 'react-icons/md'
import { BsClipboardData, BsBell } from 'react-icons/bs'
import { useNavigate } from '@reach/router';

import { ColorModeSwitcher } from "../ColorModeSwitcher"
import NewFeatureImg from '../assets/img/new.svg';
import { userLogout } from '../lib/user'

const navigationLinks = [
  {id: 1, name: '探索', 'url': 'https://www.chenxii.xyz'},
  {id: 2, name: '题库', 'url': 'https://www.chenxii.xyz'},
  {id: 3, name: '讨论', 'url': 'https://www.chenxii.xyz'},
  {id: 4, name: '竞赛', 'url': 'https://www.chenxii.xyz'},
  {id: 5, name: '企业', 'url': 'https://www.chenxii.xyz'},
]

// const newFeatureLinkImage = "https://assets.leetcode-cn.com/support/new.svg"

export const Header = () => {
  const navigationBarBg = useColorModeValue("white", "#1A202C")
  const navigationBarColor = useColorModeValue("rgba(38, 38, 38, 1)", "white")
  const navigationBarBorderBottom = useColorModeValue("rgba(240, 240, 240, 1)", "#2D3748")

  const navigate = useNavigate()
  const [alertOpen, setAlertOpen] = useState<boolean>(false)
  const cancelRef = useRef(null)

  const handleUserLogout = () => {
    userLogout()
    setAlertOpen(false)
    navigate("login")
  }

  return (
    <Box
      borderBottom={`1px solid ${navigationBarBorderBottom}`}
      fontSize="sm"
      fontWeight="450"
      color={navigationBarColor}
      bg={navigationBarBg}
      data-testid="navigationBar">
      <Box
        d="flex"
        justifyContent="space-between"
        flexWrap="nowrap"
        minW="992px"
        maxW="1200px"
        w="100%"
        mx="auto"
        px="20px"
        py="6px"
        position="relative">
        <HStack spacing="30px" data-testid="links">
          {navigationLinks.map(link => (
            <Link key={link.id} href={link.url} isExternal _hover={{textColor: 'rgba(34, 34, 34, 1)', fontWeight: 400}}>
              <Text cursor="pointer">{link.name}</Text>
            </Link>
          ))}
        </HStack>
        <HStack spacing="30px">
          <Button
            size="sm"
            leftIcon={<MdBookmarkBorder />}
            colorScheme="teal"
            variant="outline"
            cursor="pointer"
            data-testid="btn__download"
          >
            下载App
          </Button>
          <Button
            size="sm"
            leftIcon={<MdControlPoint />}
            colorScheme="yellow"
            variant="outline"
            cursor="pointer"
            data-testid="btn__premium"
          >
            Plus会员
          </Button>
          {/* FIXME: hover not working for the icon */}
          <Box _hover={{color: "blue", stroke: 'blue'}}>
            <BsClipboardData
              size={18}
              cursor="pointer"
              color="rgba(140, 140, 140, 1)"
              data-testid="icon__clipboard"
            />
          </Box>
          <Box _hover={{color: "rgba(89, 89, 89, 1)"}}>
            <BsBell
              size={18}
              cursor="pointer"
              color="rgba(140, 140, 140, 1)"
              data-testid="icon__bell"
            />
          </Box>
          {/* <Avatar size="xs" cursor="pointer" data-testid="avatar">
            <AvatarBadge boxSize="1em" bg="green.500" />
          </Avatar> */}
          <Menu>
            <MenuButton as={Button} p="0" h="24px" w="40px" bg="" _hover={{bg: ""}} _active={{bg: ""}} _focus={{boxShadow: "none"}}>
              <Avatar size="xs" cursor="pointer" data-testid="avatar">
              <AvatarBadge boxSize="1em" bg="green.500" />
            </Avatar>
            </MenuButton>
            <MenuList>
              <MenuItem onClick={() => setAlertOpen(true)}>Logout</MenuItem>
            </MenuList>
          </Menu>
          <ColorModeSwitcher/>
        </HStack> 
        <Image
          src={NewFeatureImg}
          alt="new feature"
          position="absolute"
          right="510px"
          top="-9px"
          cursor="pointer"
          transition="all 0.3s"
          _hover={{ top: "-3px" }}
        />
        <AlertDialog
          isOpen={alertOpen}
          motionPreset="slideInBottom"
          leastDestructiveRef={cancelRef}
          onClose={() => setAlertOpen(false)}
        >
          <AlertDialogOverlay>
            <AlertDialogContent>
              <AlertDialogHeader fontSize="lg" fontWeight="bold">
                Logout
              </AlertDialogHeader>

              <AlertDialogBody>
                Are you sure to logout?
              </AlertDialogBody>

              <AlertDialogFooter>
                <Button ref={cancelRef} size="sm" onClick={() => setAlertOpen(false)}>
                  Cancel
                </Button>
                <Button colorScheme="red" size="sm" onClick={handleUserLogout} ml={4}>
                  Logout
                </Button>
              </AlertDialogFooter>
            </AlertDialogContent>
          </AlertDialogOverlay>
        </AlertDialog>
      </Box>
    </Box>
  )
}
