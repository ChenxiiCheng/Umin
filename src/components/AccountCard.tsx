import React, {useState} from 'react'
import { Box, Text, Heading, InputGroup, InputLeftAddon, Input, Button, useColorModeValue, InputRightElement } from '@chakra-ui/core';
import {FaRegEnvelope, FaRegEye, FaRegEyeSlash, FaRegStar} from 'react-icons/fa'

interface IProps {
  type: 'register' | 'login'
}

export const AccountCard: React.FC<IProps> = ({ type }) => {
  const cardBgColor = useColorModeValue('white', 'rgba(255, 255, 255, 0.01)')
  const textColor = useColorModeValue('#1A202C', 'white')
  const [emailError, setEmailError] = useState(false)
  const [passwordError, setPasswordError] = useState(false)
  const [nicknameError, setNicknameError] = useState(false)
  const [showPassword, setShowPassword] = useState(false)

  return (
    <Box
      display="flex"
      flexDir="column"
      justifyContent="space-between"
      alignItems="center"
      shadow="md"
      borderWidth="1px"
      borderRadius="md"
      p={10}
      pt={9}
      minW={530}
      minH={type === 'login' ? 400 : 450}
      maxW={600}
      maxH={600}
      color={textColor}
      background={cardBgColor}
      fontSize="lg"
      fontWeight="450"
    >
      <Heading as="h2" size="xl">
        {type === "login" ? "Sign In" : "Sign Up"}
      </Heading>
      <Text color={textColor}>
        {type === "login" ? "Do not have an account?" : "Already have an account ?"}
      </Text>
      <InputGroup mt={3}>
        <InputLeftAddon
          pointerEvents="none"
          children={<FaRegEnvelope color="gray.300" />}
        />
        <Input
          type="text"
          pl="20px"
          focusBorderColor="gray.300"
          isInvalid={emailError === true ? true: false}
          errorBorderColor={emailError === true ? "red.300" : ""}
          placeholder="Email"
        />
      </InputGroup>
      {/* COMMENT: regster form has one extra input - Nickname */}
      {type === 'register' ? (
        <InputGroup>
          <InputLeftAddon
            pointerEvents="none"
            children={<FaRegStar color="gray.300" />}
          />
          <Input
            type="text"
            pl="20px"
            focusBorderColor="gray.300"
            isInvalid={nicknameError === true ? true: false}
            errorBorderColor={nicknameError === true ? "red.300" : ""}
            placeholder="Nickname"
          />
        </InputGroup>
      ) : null}
      <InputGroup>
        <InputLeftAddon
          pointerEvents="none"
          children={showPassword ? <FaRegEye color="gray.300" /> : <FaRegEyeSlash color="gray.300" />}
        />
        <Input
          type={showPassword ? "text" : "password"}
          focusBorderColor="gray.300"
          pl="20px"
          isInvalid={passwordError === true ? true: false}
          errorBorderColor={passwordError === true ? "red.300" : ""}
          placeholder="Password"
        />
        <InputRightElement>
          <Button
            h="2rem"
            w="3rem"
            size="md"
            mr={2}
            onClick={() => setShowPassword(!showPassword)}
          >
            <Text color="gray.500">{showPassword ? "Hide" : "Show"}</Text>
          </Button>
        </InputRightElement>
      </InputGroup>
      <Button color={textColor} w="100%">
        {type === "login" ? "Sign In" : "Sign Up"}
      </Button>
    </Box>
  )
}
