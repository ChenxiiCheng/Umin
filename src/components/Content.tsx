import React from 'react'
import { Box, useColorMode } from '@chakra-ui/core';

export const Content = () => {
  const { colorMode, toggleColorMode } = useColorMode()

  return (
    <Box
      mt={6}
      p={5}
      bgColor={colorMode === "light" ? "rgba(255, 255, 255, 1)" : "rgba(29, 37, 53, 1)"}
      color={colorMode === "light" ? "#1A202C" : "rgba(219, 219, 219, 1)"} fontSize="lg"
      fontWeight="450"
      boxShadow={colorMode === "light" ? "0px 1px 5px rgba(0, 0, 0, 0.1)" : "0 1px 7px rgba(0, 0, 0, 0.3)"}
    >
      hello
       Lorem ipsum dolor sit amet consectetur adipisicing elit. Optio commodi voluptate cumque, quod rerum quasi iure maxime eaque fuga voluptatibus voluptates voluptatum. Consequatur ab veritatis consectetur assumenda dignissimos in obcaecati ratione, pariatur quidem enim architecto, illo error itaque, at adipisci distinctio laboriosam impedit sed qui iure neque ipsa a! Fugiat enim, sunt odio mollitia veniam ipsum! Commodi, vel quae a nesciunt, quam temporibus possimus reprehenderit maiores veniam magni excepturi totam? Maxime minus sequi distinctio non consectetur earum harum exercitationem, nesciunt optio eaque unde? Reprehenderit esse architecto aut. Accusamus aspernatur esse nulla voluptatum dicta illo reprehenderit beatae voluptatem, sunt facilis dolore?
    </Box>
  )
}
