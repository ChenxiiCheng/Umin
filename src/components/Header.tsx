import React from 'react'
import { Text, Link, Button, HStack, Box, Avatar, AvatarBadge, Image, useColorModeValue } from '@chakra-ui/core'
import { ColorModeSwitcher } from "../ColorModeSwitcher"
import { MdBookmarkBorder, MdControlPoint } from 'react-icons/md'
import { BsClipboardData, BsBell } from 'react-icons/bs'

const navigationLinks = [
  {id: 1, name: '探索', 'url': 'https://www.chenxii.xyz'},
  {id: 2, name: '题库', 'url': 'https://www.chenxii.xyz'},
  {id: 3, name: '讨论', 'url': 'https://www.chenxii.xyz'},
  {id: 4, name: '竞赛', 'url': 'https://www.chenxii.xyz'},
  {id: 5, name: '企业', 'url': 'https://www.chenxii.xyz'},
]

const newFeatureLinkImage = "https://assets.leetcode-cn.com/support/new.svg"

export const Header = () => {
  const navigationBarBg = useColorModeValue("white", "#1A202C")
  const navigationBarColor = useColorModeValue("rgba(38, 38, 38, 1)", "white")
  const navigationBarBorderBottom = useColorModeValue("rgba(240, 240, 240, 1)", "#2D3748")

  return (
    <Box borderBottom={`1px solid ${navigationBarBorderBottom}`} fontSize="sm" fontWeight="450" color={navigationBarColor} bg={navigationBarBg} data-testid="navigationBar">
      <Box d="flex" justifyContent="space-between" flexWrap="nowrap" minW="992px" maxW="1200px" w="100%" mx="auto" px="20px" py="6px" position="relative">
        <HStack spacing="30px" data-testid="links">
          {navigationLinks.map(link => (
            <Link key={link.id} href={link.url} isExternal _hover={{textColor: 'rgba(34, 34, 34, 1)', fontWeight: 400}}>
              <Text cursor="pointer">{link.name}</Text>
            </Link>
          ))}
        </HStack>
        <HStack spacing="30px">
          <Button size="sm" leftIcon={<MdBookmarkBorder />} colorScheme="teal" variant="outline" cursor="pointer" data-testid="btn__download">下载App</Button>
          <Button size="sm" leftIcon={<MdControlPoint />} colorScheme="yellow" variant="outline" cursor="pointer" data-testid="btn__premium">Plus会员</Button>
          {/* FIXME: hover not working for the icon */}
          <Box _hover={{color: "blue", stroke: 'blue'}}>
            <BsClipboardData size={18} cursor="pointer" color="rgba(140, 140, 140, 1)" data-testid="icon__clipboard" />
          </Box>
          <Box _hover={{color: "rgba(89, 89, 89, 1)"}}>
            <BsBell size={18} cursor="pointer" color="rgba(140, 140, 140, 1)" data-testid="icon__bell" />
          </Box>
          <Avatar size="xs" cursor="pointer" data-testid="avatar"><AvatarBadge boxSize="1em" bg="green.500" /></Avatar>
          <ColorModeSwitcher/>
        </HStack> 
        <Image src={newFeatureLinkImage} alt="new feature" position="absolute" right="510px" top="-9px" cursor="pointer" transition="all 0.3s" _hover={{ top: "-3px" }} />
      </Box>
    </Box>
  )
}
