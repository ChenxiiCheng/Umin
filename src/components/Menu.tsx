import React, {useState, useEffect, useContext, createContext, Dispatch, SetStateAction} from 'react'
import { VStack, Box, HStack, useColorMode, Divider, ColorMode } from '@chakra-ui/core';
import {BsBarChart, BsTerminalFill, BsAward, BsCreditCard, BsClipboardData, BsPerson, BsShieldLock, BsBell, BsBootstrapReboot} from 'react-icons/bs'

interface IMenuItem {
  key: string
  name: string
  icon: string
}

interface IMenuItemProps {
  item: IMenuItem
}

interface IMenuListSection  {
  section: IMenuItem[]
}

interface IMenuContext {
  colorMode: ColorMode
  clickedItem: string
  setClickedItem: Dispatch<SetStateAction<string>>
}

// * Mock menu list data
const menuList: IMenuItem[][] = [
  [
    {key: 'analysis', name: '做题分析', icon: 'BsBarChart'},
    {key: 'progress', name: '进度管理', icon: 'BsTerminalFill'},
  ],
  [
    {key: 'points', name: '积分', icon: 'BsAward'},
    {key: 'coupon', name: '优惠券', icon: 'BsCreditCard'},
    {key: 'order', name: '订单', icon: 'BsClipboardData'},
  ],
  [
    {key: 'account', name: '个人资料', icon: 'BsPerson'},
    {key: 'security', name: '账号安全', icon: 'BsShieldLock'},
    {key: 'notification', name: '通知与隐私', icon: 'BsBell'},
    {key: 'plan', name: '体验计划', icon: 'BsBootstrapReboot'},
  ]
]

const VectorIcon = (props: { iconName: string }): JSX.Element => {
  switch (props.iconName) {
    case 'BsBarChart': return <BsBarChart />
    case 'BsTerminalFill': return <BsTerminalFill />
    case 'BsAward': return <BsAward />
    case 'BsCreditCard': return <BsCreditCard />
    case 'BsClipboardData': return <BsClipboardData />
    case 'BsPerson': return <BsPerson />
    case 'BsShieldLock': return <BsShieldLock />
    case 'BsBell': return <BsBell />
    case 'BsBootstrapReboot': return <BsBootstrapReboot />
    default: return <BsBarChart />
  }
}

const MenuListSection = (props: IMenuListSection) => {
  const { section } = props;

  return (
    <>
      {section && section.map((item, index) => (<MenuItem key={index} item={item} />))}
      <Divider />
    </>
  )
}

const MenuItem = (props: IMenuItemProps) => {
  const { item } = props;
  const {colorMode, clickedItem, setClickedItem} = useContext(MenuListContext)

  return (
    <Box
      as="button"
      d="flex"
      justifyContent="flex-start"
      alignItems="center"
      p={4}
      fontSize={16}
      fontWeight={450}
      mt={6}
      w="100%"
      cursor="pointer"
      borderRadius="5px"
      _hover={{ color: "rgba(45, 181, 93, 1)", bgColor: colorMode === "light" ? "rgba(0, 0, 0, 0.04)" : "rgba(255, 255, 255, 0.04)" }}
      bgColor={clickedItem === item.key ? colorMode === "light" ? "rgba(0, 0, 0, 0.04)" : "rgba(255, 255, 255, 0.04)" : ""}
      color={clickedItem === item.key ? colorMode === "light" ? "rgba(45, 181, 93, 1)" : "rgba(45, 181, 93, 1)" : "rgba(140, 140, 140, 1)"}
      outline="none"
      onClick={() => setClickedItem(item.key)}>
      <HStack spacing={3}>
        <VectorIcon iconName={item.icon} />
        <Box as="span">
          {item.name}
        </Box>
      </HStack>
    </Box>
  )
}

const MenuListContext = createContext({} as IMenuContext)
export const Menu = () => {
  const { colorMode, toggleColorMode } = useColorMode()
  const [clickedItem, setClickedItem] = useState<string>('')
  const [menuTreeNodes, setMenuTreeNodes] = useState<JSX.Element[] | null>(null)

  useEffect(() => {
    const menuTree = menuList.length !== 0 ? renderMenuList(menuList) : null
    setMenuTreeNodes(menuTree)
  }, [])

  const renderMenuList = (menuListData: IMenuItem[][]): JSX.Element[] => {
    return menuListData.map((section, sectionKey) => (<MenuListSection key={sectionKey} section={section} />))
  }

  return (
    <MenuListContext.Provider value={{colorMode, clickedItem, setClickedItem}}>
      <VStack>
        {menuTreeNodes}
      </VStack>
    </MenuListContext.Provider>
  )
}
