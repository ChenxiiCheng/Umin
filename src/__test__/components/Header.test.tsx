import React from "react"
import { cleanup } from "@testing-library/react"

import { render } from "../../test-utils"
import { Header } from '../../components/Header';

afterEach(cleanup)

describe('Test <Header />', () => {
  test('should render links, buttons, icons, avatar, image correctly', () => {
    const { getByTestId, getByAltText, container } = render(<Header />)
    
    expect(getByTestId('links')).toBeTruthy()
    expect(getByTestId('btn__download')).toBeTruthy()
    expect(getByTestId('btn__premium')).toBeTruthy()
    expect(getByTestId('icon__clipboard')).toBeTruthy()
    expect(getByTestId('icon__bell')).toBeTruthy()
    expect(getByTestId('avatar')).toBeTruthy()
    expect(getByAltText('new feature')).toBeTruthy()
    expect(container.firstChild).toMatchSnapshot()
  })
})