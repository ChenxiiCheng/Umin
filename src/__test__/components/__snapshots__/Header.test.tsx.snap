// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <Header /> should render links, buttons, icons, avatar, image correctly 1`] = `
<div
  class="css-1seza0f"
  data-testid="navigationBar"
>
  <div
    class="css-120egg3"
  >
    <div
      class="chakra-stack css-19qfvaz"
      data-testid="links"
    >
      <a
        class="chakra-link css-12b0rhz"
        href="https://www.chenxii.xyz"
        rel="noopener noreferrer"
        target="_blank"
      >
        <p
          class="chakra-text css-4g6ai3"
        >
          探索
        </p>
      </a>
      <a
        class="chakra-link css-12b0rhz"
        href="https://www.chenxii.xyz"
        rel="noopener noreferrer"
        target="_blank"
      >
        <p
          class="chakra-text css-4g6ai3"
        >
          题库
        </p>
      </a>
      <a
        class="chakra-link css-12b0rhz"
        href="https://www.chenxii.xyz"
        rel="noopener noreferrer"
        target="_blank"
      >
        <p
          class="chakra-text css-4g6ai3"
        >
          讨论
        </p>
      </a>
      <a
        class="chakra-link css-12b0rhz"
        href="https://www.chenxii.xyz"
        rel="noopener noreferrer"
        target="_blank"
      >
        <p
          class="chakra-text css-4g6ai3"
        >
          竞赛
        </p>
      </a>
      <a
        class="chakra-link css-12b0rhz"
        href="https://www.chenxii.xyz"
        rel="noopener noreferrer"
        target="_blank"
      >
        <p
          class="chakra-text css-4g6ai3"
        >
          企业
        </p>
      </a>
    </div>
    <div
      class="chakra-stack css-19qfvaz"
    >
      <button
        class="chakra-button css-7q1pq8"
        data-testid="btn__download"
        type="button"
      >
        <span
          class="chakra-button__icon css-9u48bm"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            focusable="false"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M17 3H7c-1.1 0-1.99.9-1.99 2L5 21l7-3 7 3V5c0-1.1-.9-2-2-2zm0 15l-5-2.18L7 18V5h10v13z"
            />
          </svg>
        </span>
        下载App
      </button>
      <button
        class="chakra-button css-18y46rx"
        data-testid="btn__premium"
        type="button"
      >
        <span
          class="chakra-button__icon css-9u48bm"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            focusable="false"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.49 2 2 6.49 2 12s4.49 10 10 10 10-4.49 10-10S17.51 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
            />
          </svg>
        </span>
        Plus会员
      </button>
      <div
        class="css-15gbdmn"
      >
        <svg
          color="rgba(140, 140, 140, 1)"
          cursor="pointer"
          data-testid="icon__clipboard"
          fill="currentColor"
          height="18"
          size="18"
          stroke="currentColor"
          stroke-width="0"
          style="color: rgb(140, 140, 140);"
          viewBox="0 0 16 16"
          width="18"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M4 1.5H3a2 2 0 00-2 2V14a2 2 0 002 2h10a2 2 0 002-2V3.5a2 2 0 00-2-2h-1v1h1a1 1 0 011 1V14a1 1 0 01-1 1H3a1 1 0 01-1-1V3.5a1 1 0 011-1h1v-1z"
            fill-rule="evenodd"
          />
          <path
            clip-rule="evenodd"
            d="M9.5 1h-3a.5.5 0 00-.5.5v1a.5.5 0 00.5.5h3a.5.5 0 00.5-.5v-1a.5.5 0 00-.5-.5zm-3-1A1.5 1.5 0 005 1.5v1A1.5 1.5 0 006.5 4h3A1.5 1.5 0 0011 2.5v-1A1.5 1.5 0 009.5 0h-3z"
            fill-rule="evenodd"
          />
          <path
            d="M4 11a1 1 0 112 0v1a1 1 0 11-2 0v-1zm6-4a1 1 0 112 0v5a1 1 0 11-2 0V7zM7 9a1 1 0 012 0v3a1 1 0 11-2 0V9z"
          />
        </svg>
      </div>
      <div
        class="css-1anw30t"
      >
        <svg
          color="rgba(140, 140, 140, 1)"
          cursor="pointer"
          data-testid="icon__bell"
          fill="currentColor"
          height="18"
          size="18"
          stroke="currentColor"
          stroke-width="0"
          style="color: rgb(140, 140, 140);"
          viewBox="0 0 16 16"
          width="18"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 16a2 2 0 002-2H6a2 2 0 002 2z"
          />
          <path
            clip-rule="evenodd"
            d="M8 1.918l-.797.161A4.002 4.002 0 004 6c0 .628-.134 2.197-.459 3.742-.16.767-.376 1.566-.663 2.258h10.244c-.287-.692-.502-1.49-.663-2.258C12.134 8.197 12 6.628 12 6a4.002 4.002 0 00-3.203-3.92L8 1.917zM14.22 12c.223.447.481.801.78 1H1c.299-.199.557-.553.78-1C2.68 10.2 3 6.88 3 6c0-2.42 1.72-4.44 4.005-4.901a1 1 0 111.99 0A5.002 5.002 0 0113 6c0 .88.32 4.2 1.22 6z"
            fill-rule="evenodd"
          />
        </svg>
      </div>
      <span
        class="chakra-avatar css-9rtrme"
        data-testid="avatar"
      >
        <svg
          class="css-16ite8i"
          role="img"
          viewBox="0 0 128 128"
        >
          <path
            d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z"
            fill="currentColor"
          />
          <path
            d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24"
            fill="currentColor"
          />
        </svg>
        <div
          class="chakra-avatar__badge css-1uh4lzh"
        />
      </span>
      <button
        aria-label="Switch to dark mode"
        class="chakra-button css-1jr0f57"
        data-testid="colorSwitcher"
        type="button"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          focusable="false"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M283.211 512c78.962 0 151.079-35.925 198.857-94.792 7.068-8.708-.639-21.43-11.562-19.35-124.203 23.654-238.262-71.576-238.262-196.954 0-72.222 38.662-138.635 101.498-174.394 9.686-5.512 7.25-20.197-3.756-22.23A258.156 258.156 0 0 0 283.211 0c-141.309 0-256 114.511-256 256 0 141.309 114.511 256 256 256z"
          />
        </svg>
      </button>
    </div>
    <img
      alt="new feature"
      class="chakra-image__placeholder css-1l7cmaq"
    />
  </div>
</div>
`;
